emcc.py

```diff
@@ -2451,23 +2496,48 @@ def generate_html(target, options, js_target, target_basename,
         URL.revokeObjectURL(script.src);
       };
       document.body.appendChild(script);
-    };
-    codeXHR.send(null);
-''' % (os.path.basename(asm_target), '\n'.join(asm_mods), script.inline)
+    }
+    if (fileBytes) {
+      onCodeLoad(intArrayToString(fileBytes));
+    } else {
+      codeXHR = new XMLHttpRequest();
+      codeXHR.open('GET', filename, true);
+      codeXHR.onload = function() {
+        onCodeLoad(codeXHR.responseText);
+      };
+      codeXHR.send(null);
+    }
+''' % (shared.JS.get_subresource_location(asm_target), '\n'.join(asm_mods), script.inline)
 
   if shared.Settings.BINARYEN and not shared.Settings.BINARYEN_ASYNC_COMPILATION:
     # We need to load the wasm file before anything else, it has to be synchronously ready TODO: optimize
     script.un_src()
     script.inline = '''
-          var wasmXHR = new XMLHttpRequest();
-          wasmXHR.open('GET', '%s', true);
-          wasmXHR.responseType = 'arraybuffer';
-          wasmXHR.onload = function() {
-            Module.wasmBinary = wasmXHR.response;
+          var filename = '%s';
+          var fileBytes = tryParseAsDataURI(filename);
+          function onWasmLoad() {
 %s
-          };
-          wasmXHR.send(null);
-''' % (os.path.basename(wasm_binary_target), script.inline)
+          }
+          if (fileBytes) {
+            Module.wasmBinary = fileBytes.buffer;
+            onWasmLoad();
+          } else {
+            var wasmXHR = new XMLHttpRequest();
+            wasmXHR.open('GET', filename, true);
+            wasmXHR.responseType = 'arraybuffer';
+            wasmXHR.onload = function() {
+              Module.wasmBinary = wasmXHR.response;
+              onWasmLoad();
+            };
+            wasmXHR.send(null);
+          }
+''' % (shared.JS.get_subresource_location(wasm_binary_target), script.inline)
+
+  if script.inline:
+    # add functions needed for data URI parsing
+    f = open('src/arrayUtils.js', 'r')
+    script.inline += re.sub('#if .*?#endif', '', f.read(), flags=re.DOTALL)
```

Comment:
Hmm, yeah, this is an issue. We don't run that preprocessing code on the HTML. I was hoping we would just have regular JS in this helper file, but we do have

 * that one ifdef
 * those `{{{ maybeExport }}}` things

The maybeExports are actually not needed in the html, so those could just be put in the main JS file (preamble.js unless there is some place better, with a clear comment). But that ifdef is trickier. I'll write a comment on it directly (since I want to cc @juj and that way it'll take him right there.